### PR TITLE
Endorser protocol configuration, automation and demo integration

### PR DESCRIPTION
**Pull request by ianco**
See issue https://github.com/hyperledger/aries-cloudagent-python/issues/1238

This PR adds automation support for requesting, endorsing and writing transactions, and for executing the workflow of creating a cred def that supports revocation.

The next stage of work will re-factor the code to use the event bus and eliminate the code duplication (see https://hackmd.io/51flx8CVS1ypyOjEqVl1tg?view#Implementation-Details)

### CODE REVIEW

**Comment by swcurran** [comment]
I didn't find anything in testing this using alice/faber, but I couldn't figure out how to activate it in the demo.  Do you have instructions?

We also need some tweaks to the run_demo help handling.  I'm going to put in some suggestions in a separate issue.

**Comment by ianco** [comment]
> I didn't find anything in testing this using alice/faber, but I couldn't figure out how to activate it in the demo. Do you have instructions?

The demo is in an "interim" state right now, endorser support is added but not fully added.  The following is how it works now (sets up the endorser roles to allow manual testing using the swagger page) but not ultimately how I see the demo demonstrating endorsement ...

Run the following in 2 separate shells (make sure you are running von-network and the tails server first):

```
./run_demo faber --endorser-role endorser --revocation
./run_demo alice --endorser-role author --revocation
```

Copy the invitation from faber to alice to complete the connection.

Then in the alice shell, select option "D" and copy faber's DID to alice (it is the DID displayed on faber agent startup).

This starts up the aca-py agents with the endorser role set (via the new command-line args) and sets up the connection between the 2 agents with appropriate configuration.

Then, in the alice swagger page (http://localhost:8031) you can create a schema and cred def, and all the endorser steps will happen automatically.  You don't need to specify a connection id or explicitly request endorsement (aca-py does it all automatically based on the startup args).

**Comment by swcurran** [comment]
Nice -- that worked.

I'm thinking we need to have a "von-network"-like Endorser so that we can start it when needed for testing.  Perhaps it should be built into von-network?   We should brainstorm a bit how to do that.  We also need to have one that we can put into production, ahead of a BPA-based Endorser.

I think an "Endorser.md" file is needed for this to document how to use it both in real usage, and a section on how to use it in the demo.  Seems like a fun thing to do on a Friday, no?  :-). I'm thinking that should be done before merging.

I'm hoping you can demo this on Tuesday morning at the ACA-Pug meeting.  Doable?

**Comment by ianco** [comment]
> I think an "Endorser.md" file is needed for this to document how to use it both in real usage, and a section on how to use it in the demo. Seems like a fun thing to do on a Friday, no? :-). I'm thinking that should be done before merging.

I added some docs both for ACA-Py as well as the demo.  These docs will need to be updated as the Endorser code is updated & re-factored.

File: Endorser.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Transaction Endorser Support
+
+Note that the ACA-Py transaciton support is in the process of code refactor and cleanup.  The following documents the current state, but is subject to change.
```
Comment (swcurran) [inline Endorser.md:3]: Do you anticipate the API (e.g. Admin API and Startup Parameters) changing, or are those more or less set?

Comment (ianco) [reply Endorser.md:3]: I don't anticipate any API changes but can't guarantee we won't need any

File: Endorser.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+ACA-Py supports an [Endorser Protocol](https://github.com/hyperledger/aries-rfcs/pull/586), that allows an un-privieged agent (an "Author") to request another agent (the "Endorser") to sign their transactions so they can write these transactions to the ledger.
```
Comment (swcurran) [inline Endorser.md:5]: Typo "un-privileged"

File: Endorser.md (hunk 1 above)
Comment (swcurran) [reply Endorser.md:3]: Typo "transaction"

File: Endorser.md (hunk 2 above)
Comment (swcurran) [reply Endorser.md:5]: Should have a reference to this being an Indy feature/requirement.

File: Endorser.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+Transaction Endorsement is built into the protocols for Schema, Credential Definition and Revocation, and endorsements can be explicitely requested, or ACA-Py can be configured to automate the endorsement workflow.
+
+## Setting up Connections between Authors and Endorsers
+
+Since endorsement involves message exchange between two agents, these agents must establish and configure a connection before any endorsements can be provided or requested.
+
+Once the connection is established and `active`, the "role" (either Author or Endorser) is attached to the connection using the `/transactions/{conn_id}/set-endorser-role` endpoint.  For Authors, they must additionally configure the DID of the Endorser as this is required when the Author signs the transaction (prior to sending to the Endorser for endorsement) - this is done using the `/transactions/{conn_id}/set-endorser-info` endpoint.
```
Comment (swcurran) [inline Endorser.md:13]: I think it should be possible to have the DID specified on the command line.

**Review by swcurran — APPROVED** [review]
Approved -- but you might want to clean up the typos first. Or just merge.

File: Endorser.md (hunk 3 above)
Comment (ianco) [reply Endorser.md:13]: Agree this is a command line parameter but not currently used (the auto-connection-setup support is still TBD)

**Comment by ianco** [comment]
> Approved -- but you might want to clean up the typos first. Or just merge.

I'll include the doc updates/fixes in the next PR

**Pull request merged**